pkg/agent/run.go

```diff
@@ -77,11 +78,22 @@ func Run(ctx context.Context, cfg cmds.Agent) error {
 	}
 
 	cfg.DataDir = filepath.Join(cfg.DataDir, "agent")
+	os.MkdirAll(cfg.DataDir, 0700)
 
 	if cfg.ClusterSecret != "" {
 		cfg.Token = "K10node:" + cfg.ClusterSecret
 	}
 
+	lb, err := loadbalancer.Setup(ctx, cfg)
+	if err != nil {
+		return err
+	}
+	if lb != nil {
+		if cfg.ServerURL, err = lb.LoadBalancerServerURL(); err != nil {
+			return err
+		}
+	}
```

Comment:
this looks racy as tcpproxy Start spawns a separate goroutine which calls listen func.

Reply:
It shouldn't be racy, tcpproxy Start calls `ln, err := p.netListen()("tcp", ipPort)`, `p.netListen` returns `p.ListenFunc` which we defined in loadbalancer. It starts serving using the listener with a goroutine but `lb.localAddress` which is needed by  `lb.LoadBalancerServerURL()` should be defined after the execution of `lb.proxy.Start()`. A racy condition may exist where the load balancer is not serving right away, but in the loop immediately following we call `clientaccess.AgentAccessInfoToTempKubeConfig` using that modified server url until it succeeds.

Reply:
lb.localAddress is set by lb.listen, which is called by the goroutine, isn't it?

Reply:
ah, i noticed that i misread the code. never mind!